### ISSUE TITLE
firefox-trunk package isn't latest on Linux; w/ warning: "Please restart all running instances of firefox-trunk, or you will experience problems."

### ISSUE THREAD

**Issue by stephendonner** [WPO-Foundation/wptagent#216]
Cross-reference: https://github.com/mozilla/wpt-api/issues/125

The "latest" Firefox Nightly build (which comes from, in our case, the Ubuntu Linux agent, as ```firefox-trunk```), is outdated.  See this user-agent snippet in the raw JSON[0] from a test run:

```
"User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:61.0) Gecko/20100101 Firefox/61.0 PTST/181115.161114"
```
https://launchpad.net/~ubuntu-mozilla-daily/+archive/ubuntu/ppa/+packages has the latest build, so that's correct.

I've traced the problem this far, by running as such:

```docker build --no-cache -t wptagent .```

Here's the full log, if that's a better place to start.  I've removed ```q``` from the ```apt-get install``` right before the browser packages, for extra info:

https://gist.github.com/stephendonner/040fb4bc105c969a0d38dd7a95b5138e

1) We start off strong:

```
The following NEW packages will be installed:
  adwaita-icon-theme at-spi2-core chromium-codecs-ffmpeg-extra
  dconf-gsettings-backend dconf-service firefox firefox-trunk fonts-liberation
```

2) We fetch the correct, latest ```firefox-trunk```:

```
0 upgraded, 107 newly installed, 0 to remove and 3 not upgraded.
Need to get 496 MB of archives.
After this operation, 1782 MB of additional disk space will be used.
Get:1 http://dl.google.com/linux/chrome/deb stable/main amd64 google-chrome-beta amd64 71.0.3578.53-1 [57.8 MB]
Get:2 http://ppa.launchpad.net/ubuntu-mozilla-daily/ppa/ubuntu bionic/main amd64 firefox-trunk amd64 65.0~a1~hg20181114r446158-0ubuntu0.18.04.1~umd1 [48.3 MB]
```

3) Binary swapping out (tries to) happen(s):

```
Selecting previously unselected package firefox.
Preparing to unpack .../053-firefox_63.0+build2-0ubuntu0.18.04.2_amd64.deb ...
Unpacking firefox (63.0+build2-0ubuntu0.18.04.2) ...
Selecting previously unselected package firefox-trunk.
Preparing to unpack .../054-firefox-trunk_65.0~a1~hg20181114r446158-0ubuntu0.18.04.1~umd1_amd64.deb ...
Unpacking firefox-trunk (65.0~a1~hg20181114r446158-0ubuntu0.18.04.1~umd1) ...
```

4) And now, herein lies our rub:

```
9050K .......... .......... .......... .......... .......... 98% 2.71M 0s
  9100K .......... .......... .......... .......... .......... 98% 3.02M 0s
  9150K .......... .......... .......... .......... .......... 99%  893K 0s
  9200K .......... .......... .......... .......... .......... 99% 3.69M 0s
  9250K ..........                                            100% 43.8M=6.7s

2018-11-16 05:40:34 (1.35 MB/s) - written to stdout [9482344/9482344]

Setting up libgail-common:amd64 (2.24.32-1ubuntu1) ...
Setting up humanity-icon-theme (0.6.15) ...
Setting up libgtk-3-0:amd64 (3.22.30-1ubuntu1) ...
Setting up libgtk-3-bin (3.22.30-1ubuntu1) ...
Setting up firefox (63.0+build2-0ubuntu0.18.04.2) ...
update-alternatives: using /usr/bin/firefox to provide /usr/bin/gnome-www-browser (gnome-www-browser) in auto mode
update-alternatives: using /usr/bin/firefox to provide /usr/bin/x-www-browser (x-www-browser) in auto mode
Please restart all running instances of firefox, or you will experience problems.
Setting up firefox-trunk (65.0~a1~hg20181114r446158-0ubuntu0.18.04.1~umd1) ...
Please restart all running instances of firefox-trunk, or you will experience problems.
Setting up notification-daemon (3.20.0-3) ...
Setting up opera-beta (57.0.3098.36) ...
update-alternatives: using /usr/bin/opera-beta to provide /usr/bin/x-www-browser (x-www-browser) in auto mode
update-alternatives: warning: skip creation of /usr/share/man/man1/x-www-browser.1.gz because associated file /usr/share/man/man1/opera-beta.1.gz (of link group x-www-browser) doesn't exist
update-alternatives: using /usr/bin/opera-beta to provide /usr/bin/gnome-www-browser (gnome-www-browser) in auto mode
update-alternatives: warning: skip creation of /usr/share/man/man1/gnome-www-browser.1.gz because associated file /usr/share/man/man1/opera-beta.1.gz (of link group gnome-www-browser) doesn't exist
```

[0] https://www.webpagetest.org/result/181116_GT_b4aabc81265566198048672a7ec29752/?f=json

**Comment by pmeenan** [comment]
Might be worth trying to swap the ordering and install the 2 different builds in different invocations (maybe with a killall in between in case firefox launched a background process).

Is it possible it is using the wrong path to the browser or something else is going on? The installs look like they installed 63 and 65 - where did 61 come from?

**Comment by davehunt** [comment]
Buidling the Docker image from a fresh clone of master shows it's got a recent Firefox Nightly version:

>docker run wptagent cat /usr/lib/firefox-trunk/application.ini
; This file is not used. If you modify it and want the application to use
; your modifications, move it under the browser/ subdirectory and start with
; the "-app /path/to/browser/application.ini" argument.
[App]
Vendor=Mozilla
Name=Firefox
RemotingName=firefox-trunk
CodeName=Firefox Nightly
Version=65.0a1
Profile=mozilla/firefox-trunk
BuildID=20181114144436
SourceRepository=https://hg.mozilla.org/mozilla-central
SourceStamp=073045259e75e0c8f7b8ffcd5e4bf72570f98f3e
ID={ec8030f7-c20a-464f-9b0e-13a3a9e97384}

Is it possible we have an issue with the AMI being updated, or auto-updating? I'm not familiar with how the public AMIs work...

**Comment by stephendonner** [comment]
Thx, @pmeenan; is there a way we can see / help fix the EC2 AMI image/scripts for ```wptagent```?

It's obviously crucial for us here at Mozilla to have reliable, up-to-date Firefox (and other) browsers.

**Comment by pmeenan** [comment]
If you start an instance with a SSH key and open SSH you can connect as the "ubuntu" user and screen -r to connect to the running agent screen. I was planning on taking a look but didn't get a chance today.


________________________________

**Comment by pmeenan** [comment]
I just rolled out a new set of Linux AMIs.  I'm not entirely sure why the previous one wasn't updating but I didn't spend too much time looking since it was based on 16.04. I took the opportunity to just build a new image off of 18.04.

It has all of the correct current browsers and should update fine but it's worth keeping a track of the "browser_version" metric from your tests to make sure it is updating when you release a new major version to Nightly.

The servers should pick up the new AMI's within the next hour. If you're using the auto-scaling (first, I'm sorry about how bad it is) I can't remember if it will replace existing instances or only use the new AMI ID when it scales up.
